ffi/src/lib.rs

```diff
@@ -0,0 +1,187 @@
+            } else {
+                "".into()
+            }
+        },
+        languages: {
+            let mut langs: Vec<repr_c::String> = Vec::new();
+            for language in package.summary_info().languages() {
+                langs.push(language.code().to_string().into());
+            }
+            langs.into()
+        },
+        subject: package
+            .summary_info()
+            .subject()
+            .unwrap_or_default()
+            .to_string()
+            .into(),
+        title: package
+            .summary_info()
+            .title()
+            .unwrap_or_default()
+            .to_string()
+            .into(),
+        uuid: package
+            .summary_info()
+            .uuid()
+            .unwrap_or_default()
+            .to_string()
+            .into(),
+        word_count: package.summary_info().word_count().unwrap_or_default(),
+
+        has_digital_signature: package.has_digital_signature(),
+
+        table_names: {
+            let mut names: Vec<repr_c::String> = Vec::new();
+            for table in package.tables() {
+                names.push(table.name().to_string().into());
+            }
+            names.into()
+        },
+    }
+}
+
+/// Frees the memory of the given MsiInformation.
+#[ffi_export]
+fn free_information(info: MsiInformation) {
+    drop(info);
+}
+
+/// Get the specified table from the MSI file.
+#[ffi_export]
+fn get_table(
+    path: char_p::Ref<'_>,
+    table_name: char_p::Ref<'_>,
+) -> repr_c::Vec<repr_c::Vec<repr_c::String>> {
+    let file_handle = std::fs::File::open(Path::new(path.to_str())).unwrap();
+    let mut package = Package::open(file_handle).unwrap();
+
+    let mut result: Vec<repr_c::Vec<repr_c::String>> = Vec::new();
+
+    if !package.has_table(table_name.to_str()) {
+        return Vec::new().into();
+    }
+
+    let table = package.get_table(table_name.to_str()).unwrap();
+
+    // first, we add column names
+    let mut columns: Vec<repr_c::String> = Vec::new();
+    for column in table.columns() {
+        columns.push(column.name().to_string().into());
+    }
+    result.push(columns.into());
+
+    // then, we add the rows
+    package
+        .select_rows(Select::table(table_name.to_str()))
+        .expect("select")
+        .for_each(|row| {
+            let mut row_data: Vec<repr_c::String> =
+                Vec::with_capacity(row.len());
+            for index in 0..row.len() {
+                row_data.push(row[index].to_string().into());
+            }
+            result.push(row_data.into());
+        });
+
+    result.into()
+}
+
+/// Frees the memory of the get_table result.
+#[ffi_export]
+fn free_table(table: repr_c::Vec<repr_c::Vec<repr_c::String>>) {
+    drop(table);
+}
+
+/// Generate headers/bindings for get_information() function.
+pub fn generate_headers() -> io::Result<()> {
+    ::safer_ffi::headers::builder()
+        .with_language(::safer_ffi::headers::Language::CSharp)
```

Comment:
If this `generate_headers()` function is C#-specific, should it be called `generate_csharp_headers()` or something instead?  Also wondering if the filename ought to be a parameter.